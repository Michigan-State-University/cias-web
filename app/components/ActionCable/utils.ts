import { TokenRefresher } from '@anycable/core/create-cable';

import { URIEncodeObject } from 'utils/uriUtils';
import LocalStorageService from 'utils/localStorageService';

export const createCableUrl = (
  headers: Nullable<Record<string, string>>,
): Nullable<string> => {
  if (!headers) return null;
  const { 'Access-Token': AccessToken, Uid, Client } = URIEncodeObject(headers);
  if (!AccessToken || !Uid || !Client) return null;
  return `${process.env.WEBSOCKET_URL}?access_token=${AccessToken}&uid=${Uid}&client=${Client}`;
};

export const tokenRefresher: TokenRefresher = async (transport) => {
  const url = createCableUrl(LocalStorageService.getHeaders());
  if (url) {
    transport.setURL(url);
  }
};

export const getUidFromCableUrl = (cableUrl: string) =>
  new URL(cableUrl).searchParams.get('uid');
