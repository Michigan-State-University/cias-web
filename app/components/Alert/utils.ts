import { colors, themeColors } from 'theme';

import { AlertType } from './types';

export const getBackgroundColorByType = (type: AlertType) => {
  switch (type) {
    case AlertType.WARNING: {
      return colors.chardonnay;
    }
    case AlertType.INFO:
    default: {
      return themeColors.highlight;
    }
  }
};
