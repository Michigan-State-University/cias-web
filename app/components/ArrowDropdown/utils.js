import { createContext } from 'react';

export const ArrowDropdownContext = createContext({ ref: null });
