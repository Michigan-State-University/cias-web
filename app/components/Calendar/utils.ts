import {
  MAX_VISIBLE_EVENTS_SMALL,
  MAX_VISIBLE_EVENTS_MEDIUM,
  MAX_VISIBLE_EVENTS_LARGE,
} from './constants';

export const getNumberOfEventsVisible = (calendarRows: number) => {
  switch (true) {
    case calendarRows === 5:
      return MAX_VISIBLE_EVENTS_MEDIUM;
    case calendarRows > 5:
      return MAX_VISIBLE_EVENTS_SMALL;
    case calendarRows < 5:
      return MAX_VISIBLE_EVENTS_LARGE;
    default:
      return MAX_VISIBLE_EVENTS_MEDIUM;
  }
};
