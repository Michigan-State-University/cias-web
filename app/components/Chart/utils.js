import regression from 'regression';
import range from 'lodash/range';

import { TREND_LINE_DATA_KEY } from './constants';

export const generateBarChartTrendData = (data, yKey) => {
  const trendParams = range(data.length).map(index => [
    index,
    data[index][yKey],
  ]);
  const rawTrendData = regression.polynomial(trendParams, { order: 3 });

  return rawTrendData.points.map(([, y]) => ({ [TREND_LINE_DATA_KEY]: y }));
};
