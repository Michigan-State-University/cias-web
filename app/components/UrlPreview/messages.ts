import { defineMessages } from 'react-intl';

export const scope = 'app.containers.UrlPreview';

export default defineMessages({
  422: {
    id: `${scope}.422`,
    defaultMessage: 'The provided URL is invalid',
  },
  417: {
    id: `${scope}.417`,
    defaultMessage: 'The provided URL is empty',
  },
  defaultErrorMessage: {
    id: `${scope}.defaultErrorMessage`,
    defaultMessage: 'An unknown error occurred while parsing the URL',
  },
});
