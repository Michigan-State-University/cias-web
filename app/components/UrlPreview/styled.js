import styled from 'styled-components';
import { colors } from 'theme';
import Text from 'components/Text';
import H3 from 'components/H3';

export const Thumbnail = styled.div`
  background-image: url("${props => props.image}");
  background-size: cover;
  background-color: ${colors.grey};
  background-position: center;
  width: 30%;
  border-bottom-right-radius: 10px;
  border-top-right-radius: 10px;
  &:hover{
  cursor: pointer;
  }
`;

export const Url = styled.a`
  text-decoration: none !important;
  color: ${colors.bluewood};
`;

export const ClampedText = styled(Text)`
  -webkit-line-clamp: 3;
  overflow: hidden;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -ms-box-orient: vertical;
  -o-box-orient: vertical;
  display: -webkit-box;
`;

export const ClampedTitle = styled(H3)`
  -webkit-line-clamp: 2;
  overflow: hidden;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -ms-box-orient: vertical;
  -o-box-orient: vertical;
  display: -webkit-box;
`;
