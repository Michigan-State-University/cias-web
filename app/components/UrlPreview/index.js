/**
 *
 * UrlPreview
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';

import { colors, themeColors } from 'theme';

import { useUrlMetadata } from 'utils/useUrlMetadata';

import Box from 'components/Box';
import Row from 'components/Row';
import Text from 'components/Text';
import Loader from 'components/Loader';

import { ClampedText, Thumbnail, Url, ClampedTitle } from './styled';

const UrlPreview = ({ link, handleClick }) => {
  const { metadata, isFetching } = useUrlMetadata(link);

  const redirectToLink = () => {
    if (handleClick) handleClick();
    window.open(metadata.url, '_blank');
  };

  const image = metadata?.image || metadata?.images?.[0];

  return (
    <Box my={5} width="100%">
      {metadata && (
        <Box bg={colors.white}>
          <Row>
            <Box bgc="transparent" padding={20} width={image ? '70%' : '100%'}>
              {metadata.title && (
                <Url href={metadata.url} target="_blank" onClick={handleClick}>
                  <ClampedTitle>{metadata.title}</ClampedTitle>
                </Url>
              )}
              {metadata?.description && (
                <Url href={metadata.url} target="_blank" onClick={handleClick}>
                  <ClampedText mt={8} mb={13} fontSize={11}>
                    {metadata?.description}
                  </ClampedText>
                </Url>
              )}
              {metadata.url && (
                <Url href={metadata.url} target="_blank" onClick={handleClick}>
                  <Text
                    wordBreak="break-all"
                    my={3}
                    fontSize={9}
                    color={themeColors.secondary}
                  >
                    {metadata.url}
                  </Text>
                </Url>
              )}
            </Box>
            {image && <Thumbnail image={image} onClick={redirectToLink} />}
          </Row>
        </Box>
      )}
      {isFetching && (
        <Row>
          <Loader type="inline" />
        </Row>
      )}
    </Box>
  );
};

UrlPreview.propTypes = {
  link: PropTypes.string,
  handleClick: PropTypes.func,
};

export default memo(UrlPreview);
