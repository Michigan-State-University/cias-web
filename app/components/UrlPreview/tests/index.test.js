/**
 *
 * Tests for UrlPreview
 *
 */

import React from 'react';
import 'jest-styled-components';
import { render, waitFor } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import { DEFAULT_LOCALE } from 'i18n';

import { useUrlMetadata } from 'utils/useUrlMetadata';

import UrlPreview from '../index';

describe('<UrlPreview />', () => {
  const url = 'test-url';
  const props = { link: url };
  const testResponse = {
    url,
    title: 'Test',
    siteName: 'Test',
    description: 'Test site - mock data',
    images: ['image1.png'],
    mediaType: 'image.other',
    contentType: 'text/html; charset=utf-8',
    videos: [],
    favicons: ['favicon1.ico'],
  };

  beforeAll(() => {
    useUrlMetadata.mockImplementationOnce(() => ({
      metadata: testResponse,
      isFetching: false,
    }));
  });

  it('Should render and match the snapshot', async () => {
    const { container } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <UrlPreview {...props} />
      </IntlProvider>,
    );

    expect(container).toMatchSnapshot();
    await waitFor(() => expect(useUrlMetadata).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(useUrlMetadata).toHaveBeenCalledWith(url));
  });
});
