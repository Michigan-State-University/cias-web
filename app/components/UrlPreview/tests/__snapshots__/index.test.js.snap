// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UrlPreview /> Should render and match the snapshot 1`] = `
.c0 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  width: 100%;
  margin-top: 5px;
  margin-bottom: 5px;
  color: unset;
}

.c1 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  background-color: rgba(255,255,255,1);
  color: unset;
}

.c3 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  width: 70%;
  padding: 20px;
  color: unset;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
}

.c7 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-top: 8px;
  margin-bottom: 13px;
  font-size: 11px;
  color: #2F3850;
}

.c9 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-top: 3px;
  margin-bottom: 3px;
  font-size: 9px;
  color: rgba(183,39,234,1);
  word-break: break-all;
  color: rgba(183,39,234,1);
}

.c5 {
  margin: 0;
  font-family: DM Sans;
  font-size: 15px;
  line-height: 150%;
  color: #2F3850;
}

.c10 {
  background-image: url('image1.png');
  background-size: cover;
  background-color: #9AA1B6;
  background-position: center;
  width: 30%;
  border-bottom-right-radius: 10px;
  border-top-right-radius: 10px;
}

.c10:hover {
  cursor: pointer;
}

.c4 {
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #2F3850;
}

.c8 {
  -webkit-line-clamp: 3;
  overflow: hidden;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -ms-box-orient: vertical;
  -o-box-orient: vertical;
  display: -webkit-box;
}

.c6 {
  -webkit-line-clamp: 2;
  overflow: hidden;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -ms-box-orient: vertical;
  -o-box-orient: vertical;
  display: -webkit-box;
}

<div>
  <div
    class="c0"
    width="100%"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          width="70%"
        >
          <a
            class="c4"
            href="test-url"
            target="_blank"
          >
            <h3
              class="c5 c6"
            >
              Test
            </h3>
          </a>
          <a
            class="c4"
            href="test-url"
            target="_blank"
          >
            <p
              class="c7 c8"
              font-size="11"
            >
              Test site - mock data
            </p>
          </a>
          <a
            class="c4"
            href="test-url"
            target="_blank"
          >
            <p
              class="c9"
              color="#B727EA"
              font-size="9"
            >
              test-url
            </p>
          </a>
        </div>
        <div
          class="c10"
        />
      </div>
    </div>
  </div>
</div>
`;
